In [1]:
import pandas as pd
import matplotlib.pyplot as plt
from matplotlib.ticker import StrMethodFormatter

from src.config import DATA
from src.utils import age_calculator, cities_comprobation


In [2]:
df=pd.read_csv(DATA)

/tmp/ipykernel_60280/2602395650.py:1: DtypeWarning: Columns (2) have mixed types. Specify dtype option on import or set low_memory=False.
  df=pd.read_csv(DATA)


In [3]:
df.head()

,Region,Country,State,City,Month,Day,Year,AvgTemperature
0,Africa,Algeria,NaN,Algiers,1,1,1995,64.2
1,Africa,Algeria,NaN,Algiers,1,2,1995,49.4
2,Africa,Algeria,NaN,Algiers,1,3,1995,48.8
3,Africa,Algeria,NaN,Algiers,1,4,1995,46.4
4,Africa,Algeria,NaN,Algiers,1,5,1995,47.9


In [4]:
df.shape

(2906327, 8)

In [5]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2906327 entries, 0 to 2906326
Data columns (total 8 columns):
 #   Column          Dtype  
---  ------          -----  
 0   Region          object 
 1   Country         object 
 2   State           object 
 3   City            object 
 4   Month           int64  
 5   Day             int64  
 6   Year            int64  
 7   AvgTemperature  float64
dtypes: float64(1), int64(3), object(4)
memory usage: 177.4+ MB


In [6]:
df.isnull().sum()

Region                  0
Country                 0
State             1450990
City                    0
Month                   0
Day                     0
Year                    0
AvgTemperature          0
dtype: int64

In [7]:
df["Region"].nunique()

7

In [8]:
pd.DataFrame(df["Region"].value_counts())

,count
Region,
North America,1556681
Europe,381990
Asia,316663
Africa,251118
South/Central America & Carribean,219530
Middle East,124749
Australia/South Pacific,55596


In [9]:
df["Country"].nunique()

125

In [10]:
pd.DataFrame(df["Country"].value_counts())

,count
Country,
US,1455337
Canada,74245
Australia,46330
China,46329
India,37063
...,...
Guyana,5065
Israel,4641
Burundi,4543


In [11]:
df["City"].nunique()

321

In [12]:
df_city=pd.DataFrame(df["City"].value_counts())

In [13]:
df_city[df_city["count"]==18530]

,count
City,
Springfield,18530
Columbus,18530
Portland,18530
Charleston,18530
Washington DC,18530
Washington,18530


In [14]:
df_country=df[df["Country"]=="US"]

In [15]:
df_country.head()

,Region,Country,State,City,Month,Day,Year,AvgTemperature
1450990,North America,US,Alabama,Birmingham,1,1,1995,50.7
1450991,North America,US,Alabama,Birmingham,1,2,1995,37.2
1450992,North America,US,Alabama,Birmingham,1,3,1995,33.2
1450993,North America,US,Alabama,Birmingham,1,4,1995,33.3
1450994,North America,US,Alabama,Birmingham,1,5,1995,26.4


In [16]:
df_country.isnull().sum()

Region            0
Country           0
State             0
City              0
Month             0
Day               0
Year              0
AvgTemperature    0
dtype: int64

In [17]:
df_election=df[df["City"].isin(["Springfield","Columbus","Portland","Charleston","Washington DC","Washington"])]

In [18]:
df_election["Year"].value_counts()

Year
1996    4392
2000    4392
2004    4392
2016    4392
2012    4392
2008    4392
1999    4380
1995    4380
1998    4380
1997    4380
2003    4380
2002    4380
2001    4380
2006    4380
2007    4380
2009    4380
2010    4380
2005    4380
2011    4380
2013    4380
2014    4380
2015    4380
2017    4380
2018    4380
2019    4380
2020    1608
Name: count, dtype: int64

In [19]:
df_election["City"].value_counts()

City
Washington       18530
Washington DC    18530
Columbus         18530
Springfield      18530
Portland         18530
Charleston       18530
Name: count, dtype: int64

In [20]:
years=df_election["Year"].unique()

In [21]:
age_calculator(years,df_election)

Year  City         
1995  Charleston       730
      Columbus         730
      Portland         730
      Springfield      730
      Washington       730
      Washington DC    730
Name: count, dtype: int64
Year  City         
1996  Charleston       732
      Columbus         732
      Portland         732
      Springfield      732
      Washington       732
      Washington DC    732
Name: count, dtype: int64
Year  City         
1997  Charleston       730
      Columbus         730
      Portland         730
      Springfield      730
      Washington       730
      Washington DC    730
Name: count, dtype: int64
Year  City         
1998  Charleston       730
      Columbus         730
      Portland         730
      Springfield      730
      Washington       730
      Washington DC    730
Name: count, dtype: int64
Year  City         
1999  Charleston       730
      Columbus         730
      Portland         730
      Springfield      730
      Washington       730
      Washing

In [22]:
df_copy= df_election[df_election["City"]=="Columbus"]

df_copy["State"].value_counts()

State
Georgia    9265
Ohio       9265
Name: count, dtype: int64

In [23]:
cities= ["Springfield","Columbus","Portland","Charleston","Washington DC","Washington"]

In [24]:
cities_comprobation(cities,df_election)

NAME : Springfield
State
Illinois    9265
Missouri    9265
Name: count, dtype: int64
----------------------------------------
NAME : Columbus
State
Georgia    9265
Ohio       9265
Name: count, dtype: int64
----------------------------------------
NAME : Portland
State
Maine     9265
Oregon    9265
Name: count, dtype: int64
----------------------------------------
NAME : Charleston
State
South Carolina    9265
West Virginia     9265
Name: count, dtype: int64
----------------------------------------
NAME : Washington DC
State
Maryland    18530
Name: count, dtype: int64
----------------------------------------
NAME : Washington
State
District of Columbia    18530
Name: count, dtype: int64
----------------------------------------


In [25]:
df_washington=df_election[df_election["City"]=="Washington DC"]

In [26]:
df_washington[df_washington.duplicated()]

,Region,Country,State,City,Month,Day,Year,AvgTemperature
1696320,North America,US,Maryland,Washington DC,1,1,1995,40.6
1696321,North America,US,Maryland,Washington DC,1,2,1995,39.8
1696322,North America,US,Maryland,Washington DC,1,3,1995,29.3
1696323,North America,US,Maryland,Washington DC,1,4,1995,33.0
1696324,North America,US,Maryland,Washington DC,1,5,1995,20.9
...,...,...,...,...,...,...,...,...
1705580,North America,US,Maryland,Washington DC,5,9,2020,45.2
1705581,North America,US,Maryland,Washington DC,5,10,2020,50.2
1705582,North America,US,Maryland,Washington DC,5,11,2020,56.1
1705583,North America,US,Maryland,Washington DC,5,12,2020,52.5


In [27]:
df=df_election[(df_election["City"] == "Columbus") & (df_election["State"] == "Ohio") & (df_election["Year"] != 2020)].copy()

In [28]:
df.head()

,Region,Country,State,City,Month,Day,Year,AvgTemperature
2402743,North America,US,Ohio,Columbus,1,1,1995,42.4
2402744,North America,US,Ohio,Columbus,1,2,1995,21.4
2402745,North America,US,Ohio,Columbus,1,3,1995,22.5
2402746,North America,US,Ohio,Columbus,1,4,1995,14.3
2402747,North America,US,Ohio,Columbus,1,5,1995,8.1


In [29]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 9131 entries, 2402743 to 2411873
Data columns (total 8 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   Region          9131 non-null   object 
 1   Country         9131 non-null   object 
 2   State           9131 non-null   object 
 3   City            9131 non-null   object 
 4   Month           9131 non-null   int64  
 5   Day             9131 non-null   int64  
 6   Year            9131 non-null   int64  
 7   AvgTemperature  9131 non-null   float64
dtypes: float64(1), int64(3), object(4)
memory usage: 642.0+ KB


In [30]:
df.groupby("Year")["City"].value_counts()

Year  City    
1995  Columbus    365
1996  Columbus    366
1997  Columbus    365
1998  Columbus    365
1999  Columbus    365
2000  Columbus    366
2001  Columbus    365
2002  Columbus    365
2003  Columbus    365
2004  Columbus    366
2005  Columbus    365
2006  Columbus    365
2007  Columbus    365
2008  Columbus    366
2009  Columbus    365
2010  Columbus    365
2011  Columbus    365
2012  Columbus    366
2013  Columbus    365
2014  Columbus    365
2015  Columbus    365
2016  Columbus    366
2017  Columbus    365
2018  Columbus    365
2019  Columbus    365
Name: count, dtype: int64

In [31]:
df["AvgTemperature"].describe()

count    9131.000000
mean       53.590056
std        19.403705
min       -99.000000
25%        38.750000
50%        55.700000
75%        70.300000
max        89.600000
Name: AvgTemperature, dtype: float64

In [32]:
df[df["AvgTemperature"] < 0]

,Region,Country,State,City,Month,Day,Year,AvgTemperature
2404196,North America,US,Ohio,Columbus,12,24,1998,-99.0
2404197,North America,US,Ohio,Columbus,12,25,1998,-99.0
2404202,North America,US,Ohio,Columbus,12,30,1998,-99.0
2404203,North America,US,Ohio,Columbus,12,31,1998,-99.0
2404213,North America,US,Ohio,Columbus,1,10,1999,-99.0
2405468,North America,US,Ohio,Columbus,6,18,2002,-99.0
2405469,North America,US,Ohio,Columbus,6,19,2002,-99.0
2405470,North America,US,Ohio,Columbus,6,20,2002,-99.0
2405471,North America,US,Ohio,Columbus,6,21,2002,-99.0
2405549,North America,US,Ohio,Columbus,9,7,2002,-99.0


In [33]:
df[df["AvgTemperature"] == -99].count()


Region            14
Country           14
State             14
City              14
Month             14
Day               14
Year              14
AvgTemperature    14
dtype: int64

In [34]:
(df["AvgTemperature"] == -99).sum()

np.int64(14)

In [35]:
df_temperature=df[df["AvgTemperature"] == -99]

In [36]:
temperaturas=[24, 21.0, 18.0, 19.0, 29.0, 70.0, 71.0, 73.06, 73.0, 72.0, 40.0, 73.06, 73.0, 45.0]

In [37]:
df_temperature.head(20)

,Region,Country,State,City,Month,Day,Year,AvgTemperature
2404196,North America,US,Ohio,Columbus,12,24,1998,-99.0
2404197,North America,US,Ohio,Columbus,12,25,1998,-99.0
2404202,North America,US,Ohio,Columbus,12,30,1998,-99.0
2404203,North America,US,Ohio,Columbus,12,31,1998,-99.0
2404213,North America,US,Ohio,Columbus,1,10,1999,-99.0
2405468,North America,US,Ohio,Columbus,6,18,2002,-99.0
2405469,North America,US,Ohio,Columbus,6,19,2002,-99.0
2405470,North America,US,Ohio,Columbus,6,20,2002,-99.0
2405471,North America,US,Ohio,Columbus,6,21,2002,-99.0
2405549,North America,US,Ohio,Columbus,9,7,2002,-99.0


In [38]:
df_temperature.loc[df_temperature["AvgTemperature"] == -99, "AvgTemperature"] = temperaturas

In [39]:
df_temperature.head(20)

,Region,Country,State,City,Month,Day,Year,AvgTemperature
2404196,North America,US,Ohio,Columbus,12,24,1998,24.00
2404197,North America,US,Ohio,Columbus,12,25,1998,21.00
2404202,North America,US,Ohio,Columbus,12,30,1998,18.00
2404203,North America,US,Ohio,Columbus,12,31,1998,19.00
2404213,North America,US,Ohio,Columbus,1,10,1999,29.00
2405468,North America,US,Ohio,Columbus,6,18,2002,70.00
2405469,North America,US,Ohio,Columbus,6,19,2002,71.00
2405470,North America,US,Ohio,Columbus,6,20,2002,73.06
2405471,North America,US,Ohio,Columbus,6,21,2002,73.00
2405549,North America,US,Ohio,Columbus,9,7,2002,72.00


In [40]:
df_temperature.index

Index([2404196, 2404197, 2404202, 2404203, 2404213, 2405468, 2405469, 2405470,
       2405471, 2405549, 2405724, 2407365, 2407758, 2407955],
      dtype='int64')

In [41]:
df.loc[df_temperature.index, ["Year", "Month", "Day", "AvgTemperature"]]

,Year,Month,Day,AvgTemperature
2404196,1998,12,24,-99.0
2404197,1998,12,25,-99.0
2404202,1998,12,30,-99.0
2404203,1998,12,31,-99.0
2404213,1999,1,10,-99.0
2405468,2002,6,18,-99.0
2405469,2002,6,19,-99.0
2405470,2002,6,20,-99.0
2405471,2002,6,21,-99.0
2405549,2002,9,7,-99.0


In [42]:
df.loc[df_temperature.index, "AvgTemperature"]= df_temperature["AvgTemperature"]

In [43]:
(df["AvgTemperature"] ==-99).sum()

np.int64(0)

In [44]:
df.loc[df_temperature.index, ["Year", "Month", "Day", "AvgTemperature"]]

,Year,Month,Day,AvgTemperature
2404196,1998,12,24,24.00
2404197,1998,12,25,21.00
2404202,1998,12,30,18.00
2404203,1998,12,31,19.00
2404213,1999,1,10,29.00
2405468,2002,6,18,70.00
2405469,2002,6,19,71.00
2405470,2002,6,20,73.06
2405471,2002,6,21,73.00
2405549,2002,9,7,72.00


In [45]:
df["AvgTemperature"].describe()

count    9131.000000
mean       53.818631
std        18.482155
min        -3.800000
25%        38.800000
50%        55.800000
75%        70.400000
max        89.600000
Name: AvgTemperature, dtype: float64

In [46]:
df["Date"]= pd.to_datetime(df[["Year", "Month", "Day"]])

In [47]:
df[["Year", "Month", "Day", "Date"]].head()

,Year,Month,Day,Date
2402743,1995,1,1,1995-01-01
2402744,1995,1,2,1995-01-02
2402745,1995,1,3,1995-01-03
2402746,1995,1,4,1995-01-04
2402747,1995,1,5,1995-01-05


In [48]:
df["Date"].dtype

dtype('<M8[ns]')

In [49]:
df["Date"].is_monotonic_increasing

True

In [50]:
df["Date"].duplicated().sum()

np.int64(0)

In [51]:
df["Date"].min()

Timestamp('1995-01-01 00:00:00')

In [52]:
df["Date"].max()

Timestamp('2019-12-31 00:00:00')

In [53]:
expected_dates= pd.date_range(
    start=df["Date"].min(),
    end= df["Date"].max(),
    freq="D"
)

In [54]:
print(f"expected_dates : {len(expected_dates)}")
print(f"Dates Data frame: {len(df)}")



expected_dates : 9131
Dates Data frame: 9131


In [55]:
df_monthly_avg=pd.DataFrame(df.groupby(["Year","Month"])["AvgTemperature"].mean().round(2))

In [56]:
df_monthly_avg.head(24)

AvgTemperature
Year Month                
1995 1               29.89
     2               28.15
     3               44.34
     4               50.57
     5               60.98
     6               72.39
     7               75.59
     8               76.94
     9               63.60
     10              55.74
     11              38.13
     12              29.20
1996 1               27.93
     2               30.38
     3               35.52
     4               50.31
     5               60.99
     6               71.47
     7               73.51
     8               73.67
     9               65.41
     10              55.00
     11              37.28
     12              37.36

In [57]:
index_min= df_monthly_avg["AvgTemperature"].idxmin()
temperature_min=df_monthly_avg["AvgTemperature"].min()

index_max= df_monthly_avg["AvgTemperature"].idxmax()
temperature_max=df_monthly_avg["AvgTemperature"].max()

print(f"The month-year with the lowest average temperature: {index_min[0]}/{index_min[1]}")
print(f"the temperature was: {temperature_min}")
print("-"*60)
print(f"The month-year with the highest average temperature: {index_max[0]}/{index_max[1]}")
print(f"the temperature was: {temperature_max}")

The month-year with the lowest average temperature: 2015/2
the temperature was: 19.36
------------------------------------------------------------
The month-year with the highest average temperature: 2012/7
the temperature was: 80.25


### Monthly Analysis Conclusion

Between 1995 and 2019, the monthly average temperatures in Columbus show a clear seasonal pattern, with lower temperatures during winter and higher temperatures during summer.

The lowest monthly average temperature was recorded in **February 2015, at 19.36 °F**, while the highest monthly average temperature occurred in **July 2012, at 80.25 °F**.

These results show a significant variation in temperature throughout the year.

In [58]:
df.head()

,Region,Country,State,City,Month,Day,Year,AvgTemperature,Date
2402743,North America,US,Ohio,Columbus,1,1,1995,42.4,1995-01-01
2402744,North America,US,Ohio,Columbus,1,2,1995,21.4,1995-01-02
2402745,North America,US,Ohio,Columbus,1,3,1995,22.5,1995-01-03
2402746,North America,US,Ohio,Columbus,1,4,1995,14.3,1995-01-04
2402747,North America,US,Ohio,Columbus,1,5,1995,8.1,1995-01-05


In [59]:
df= df.drop(columns=["Year", "Month", "Day"])

In [60]:
df.head()

,Region,Country,State,City,AvgTemperature,Date
2402743,North America,US,Ohio,Columbus,42.4,1995-01-01
2402744,North America,US,Ohio,Columbus,21.4,1995-01-02
2402745,North America,US,Ohio,Columbus,22.5,1995-01-03
2402746,North America,US,Ohio,Columbus,14.3,1995-01-04
2402747,North America,US,Ohio,Columbus,8.1,1995-01-05


In [79]:
index_lowest_temperature_day= df["AvgTemperature"].idxmin()
coldest_day= df.loc[index_lowest_temperature_day, "Date"].strftime("%Y-%m-%d")
lowest_temperature= df["AvgTemperature"].min()

index_highest_temperature_day= df["AvgTemperature"].idxmax()
hottest_day=df.loc[index_highest_temperature_day, "Date"].strftime("%Y-%m-%d")
highest_temperature= df["AvgTemperature"].max()

print(f"The coldest day was: {coldest_day}")
print(f"The temperature was : {lowest_temperature} °F")
print("-"*60)
print(f"the hottest day was: {hottest_day}")
print(f"The temperature was : {highest_temperature} °F")

The coldest day was: 2009-01-16
The temperature was : -3.8 °F
------------------------------------------------------------
the hottest day was: 2012-07-07
The temperature was : 89.6 °F


### Daily Temperature Extremes Conclusion

Between 1995 and 2019, the coldest day recorded in Columbus was **January 16, 2009**, with an average temperature of **-3.8 °F**.

The hottest day was **July 7, 2012**, with an average temperature of **89.6 °F**.

These results show the wide range of daily temperatures experienced in Columbus over the analyzed period.